### PR TITLE
Reintegrate external profiling support

### PR DESCRIPTION
**Pull request by fisiognomico**
# Julia external profiling plus perf support

As presented in #26999, during ORC JIT transition Julia lost the LLVM native support to register JIT events. This was my first activity.

Also with my mentors of the Gsoc we agreed on adding perf support, similarly to what @maleadt did in 
#14727 .

Then I have added IntelJITEventListener and OProfile support. OProfile is currently missing line level code annotations, as the current OProfile LLVM support is not implementing it. 

That's why I have linked directly the OPprofileJITEventListener code, to avoid having to re-compile LLVM each time :sweat_smile: .

This is an example of visualizing the Perf output of runtime code using [gprof2dot](https://github.com/jrfonseca/gprof2dot).

output

## TODO 
- [x] Work on making an LLVM Patch to add DWARF debug symbols support in Operf.
- [x] After that, the OProfileJITEventListener source can be ditched, and the LLVM event listener can be registered directly to the Julia execution engine.
- [x] Maybe move the `PerfJITEventListener` class somewhere else, like in `debuginfo.cpp`, as it is quite awful linked in the Makefile.
 @vtjnash suggested to add some features to the `@profile` macro. This can be useful, as perf support is intended to profile the whole runtime. I can add a feature so that a particular section of Julia code is profiled with perf events. Also it would not require to compile Julia with perf support.
If anybody fthinks that this can be useful, I will work on that too.

EDIT: The perf wrapper is maintained as an external package.

cc @vtjnash @vchuravy

### CODE REVIEW

File: deps/checksums/llvm-6.0.0.src.tar.xz/md5
Comment (vchuravy) [inline deps/checksums/llvm-6.0.0.src.tar.xz/md5]: Maybe rebase onto master which already should have these files (You should not check them int)

File: src/OProfileJITEventListener.cpp
Comment (vchuravy) [inline src/OProfileJITEventListener.cpp]: Formatting seems weird here

**Review by vchuravy — COMMENTED** [review]
I would still prefer if we could use the `OProfileWrapper` from LLVM instead of duplicating it here. Are there any patches we need for that?

**Comment by DokFaust** [comment]
Well if we want to integrate line level annotations for OProfile yup, I need to patch the LLVM integration. 

Either way the OProfile code will be ditched as it is redundant, and I will register the event from LLVM, as is done with `IntelJITEventListener`.
I have added it as for the moment is more convenient to test than recompiling every time LLVM :smile:

File: deps/llvm.mk
Comment (vchuravy) [inline deps/llvm.mk]: You should probably rebase your branch onto current master so that you have the other 6.0 patches as well

**Review by vchuravy — COMMENTED** [review]
Great! `PerfJITEventListener.cpp` is from LLVM as well right? If so we should also add this as a patch and any changes that you made should be contributed back upstream.

**Comment by DokFaust** [comment]
Ehm this rebase went too far I think. I am working on fixin this

**Comment by vchuravy** [comment]
You probably also want to squash your commits into one introducing the LLVM patches and one or two for the integration work.

**Comment by DokFaust** [comment]
Yup it should be cleaner now
Also the OProfile patch has now been upstreamed!

As for a wrapper to Perf events in Julia [LinuxPerf.jl](https://github.com/DokFaust/LinuxPerf.jl) will be maintained as an external package, so I took that out of th todo list.

File: src/jitlayers.cpp
Comment (vtjnash) [inline src/jitlayers.cpp]: space between `if` and `(`

File: src/jitlayers.cpp
Comment (vtjnash) [inline src/jitlayers.cpp]: space between `for` and `(`

File: src/codegen.cpp
Comment (vtjnash) [inline src/codegen.cpp]: space between `if` and `(`

**Review by vtjnash — COMMENTED** [review]
Looks good to me (other than some minor formatting changes). Can you rebase to get CI to re-run? We'll also need BinaryBuilder to ingest these patches. But since this is off by default, I don't think that needs happen before merging.

**Review by vtjnash — APPROVED** [review]

**Comment by vtjnash** [comment]
(CI failures are unrelated – httpbin.com flakiness and macOS FileWatching)

File: deps/llvm.mk
Comment (vchuravy) [inline deps/llvm.mk]: Duplicate

**Review by vchuravy — CHANGES_REQUESTED** [review]
Please squash the commits and do we need to add https://reviews.llvm.org/D47343 and https://reviews.llvm.org/D44890?

**Comment by DokFaust** [comment]
`D44890` and `D47343` are superfluous as we do object notification in JuliaOJIT!

**Review by vchuravy — APPROVED** [review]
LGTM, after CI passes

**Pull request merged**
